### ISSUE TITLE
(🎁) Support continuable failures in python code

### ISSUE THREAD

**Issue by KotlinIsland** [robotframework/robotframework#4276]
Desired functionality: perform "run keyword and continue on failure" and support continuable failures in python code

I have many instances where I want to report multiple failures in python code, (and want the code to flow naturally). I have created all sorts of custom implementations to get close to the desired behavior.

```robot
Continue Example
    Validate All Things On Page
    Log    end
```
(psuedo code moment)
Introduce a function that will log a failure to the log, but will also affect the status of the test:
```py
def validate_all_things_on_page():
  continuable_validate("expected a", get_a())
  continuable_validate("expected b", get_b())
  continuable_validate("expected c", get_c())
  print("Done!")

def continuable_validate(expected, actual):
    if expected != actual:
        robot.api.fail("Unfortunately a '!=' moment was detected in your code 😳", continue=True)
```

An alternate(or additional) solution would be to introduce a context manager that controls the state of failures being reported:
```py
def validate_all_things_on_page():
  with robot.api.continue_on_failures():
    validate("expected a", get_a())
    validate("expected b", get_b())
    validate("expected c", get_c())
  print("Done!")

def validate(expected, actual):
    if expected != actual:
        robot.api.fail("Unfortunately a '!=' moment was detected in your code 😳")
```
![image](https://user-images.githubusercontent.com/65446343/158305226-bad46789-b29a-48d5-8f6e-6b51a67d5df2.png)

(from https://github.com/robotframework/robotframework/issues/4237#issuecomment-1067549082)

> Being able to log multiple individual failure messages under a single keyword could be nice, though. That we could easily support by making it possible to raise an exception so that its message is actually a list of messages. In the simplest form it could just mean `AssertiongError(['First error', 'Second error'])`. We could log each error separately and then generate "Several failures occurred" error based on the errors automatically. This would be easy to implement and doesn't have that much to do with an API to run other keywords. It would need its own issue, though, so please submit one!
>- @pekkaklarck 

Duplicate of: #3706 #4038

**Comment by pekkaklarck** [comment]
If I understand this correctly, the idea is to report multiple errors from one keyword so that
- each error is logged separately and
- Robot constructs the final "Multiple failures occurred" message automatically.

If the above is the use case, what's proposed looks _really_ complicated to use and to implement. I don't understand why this would require adding both `continue_on_failures´ context manager an new `robot.api.fail` method. Could you are both needed or what the API should be?

**Comment by pekkaklarck** [comment]
As you probably already noticed, I submitted #4277 about an alternative way to report multiple failures. That approach would be easier to use because you would keep using `raise AssertionError` like earlier an it ought to be pretty easy to implement as well. The only downside I see compared to what's proposed here is that it doesn't, in the current form, support accurate timestamps with each failure.

**Comment by KotlinIsland** [comment]
> I don't understand why this would require adding both `continue_on_failures´ context manager an new `robot.api.fail` method.

I'm just providing possible solutions, I have custom functionality for both approaches.


> it doesn't, in the current form, support accurate timestamps with each failure.

I think that preserving log order is highly important. also I can see that your approach would involve heavy boilerplate with high potential for inadvertently introducing bugs.

**Comment by pekkaklarck** [comment]
Do you mean both `continue_on_failures` and `robot.api.fail` would be need to accomplish this or that they would provide two alternative ways to do that? I value your proposals, but I would very highly appreciate if you could spend a bit long time describing them so that it would be easier for me to understand them.

**Comment by KotlinIsland** [comment]
Yeah, two different approaches. Sorry about that, I'll update the OP

**Comment by pekkaklarck** [comment]
Do you want both or are you just proposing two alternatives?

**Comment by KotlinIsland** [comment]
I have both and tend to use the context manager solution a lot more, but it's entirely up to you 😀.

**Comment by KotlinIsland** [comment]
And maybe my ideas aren't optimal, but hopefully they add to the general discussion.

**Comment by pekkaklarck** [comment]
So the main idea is adding `robot.api.fail()` that can be called with `continue=True` to make the failure continuable? And then `robot.api.continue_on_failures` context manager would just change the default behavior of `robot.api.fail`?

**Comment by pekkaklarck** [comment]
Do you have implementation somewhere so that I could look at what you've actually done? I cannot think any really clean way how to implement this exact API. I do have an idea about a similar API that could probably be implemented pretty easily and cleanly.

**Comment by pekkaklarck** [comment]
Here's an example of using an API that ought to be pretty easy to implement:

```python
from robot.api import logger

def kw(expected, actual):
    with logger.failure_reporter() as reporter:
        for exp, act in zip(expected, actual):
            if exp != act:
                reporter.fail(f'{exp} != {act}')
```

The idea is that the `failure_reporter` context manager would support reporting failures so that they are handled immediately. In its `__exit__` it would then check were any failures reported and raise an exception if here were. The context manager itself would be easy to implement otherwise, but we needed to expose`fail` so that it could be used by libraries.

This API could be even nice to use:
```python
from robot.api import logger

def kw(expected, actual):
    with logger.continue_on_failure:
        for exp, act in zip(expected, actual):
            if exp != act:
                logger.fail(f'{exp} != {act}')
```
Now `logger` itself would get `fail` function. It could just raise AssertionError by default, but `continue_on_failure´ context manager would change the behavior. The annoying thing here is that because `logger` is a module, this approach would require storing some information in module global state. That's not nice in general and it would cause problems with threads.
